lib/internal/inspector_node_domains.js

```diff
@@ -0,0 +1,433 @@
+    }
+    if (this._discoverTargets) {
+      const targetInfo = this._toProtocolTarget(target);
+      this._notificationCallback('targetDestroyed', { targetInfo });
+    }
+  }
+
+  _onSessionAttached(targetId, sessionId, callback, error, waitingForDebugger,
+                     messageCallback, detachCallback) {
+    if (error) {
+      callback(error);
+      return;
+    }
+    const session = { targetId, detachCallback, messageCallback };
+    this._sessions.set(sessionId, session);
+    const targetInfo = this._toProtocolTarget(targetsAgent.getTarget(targetId));
+    if (this._sessionsForTarget(targetId).length === 1)
+      this._sendTargetInfoChanged(targetInfo);
+    this._notificationCallback('attachedToTarget',
+                               { targetInfo, sessionId, waitingForDebugger });
+    callback(null, { sessionId });
+  }
+
+  _sendTargetInfoChanged(targetInfo) {
+    this._notificationCallback('targetInfoChanged', { targetInfo });
+  }
+
+  _sessionDetached(sessionId, target) {
+    if (!this._sessions.has(sessionId))
+      return;
+    const { targetId } = this._sessions.get(sessionId);
+    this._sessions.delete(sessionId);
+    this._notificationCallback('detachedFromTarget', { sessionId });
+    if (this._sessionsForTarget(targetId).length === 0) {
+      this._sendTargetInfoChanged(this._toProtocolTarget(target));
+    }
+  }
+
+  _removeListeners() {
+    for (const remover of this._listeners) {
+      remover();
+    }
+  }
+
+  _forwardEventAsNotification(event, wrap) {
+    const listener = (data) => {
+      const targetInfo = this._toProtocolTarget(data);
+      const object = wrap ? { targetInfo } : targetInfo;
+      this._notificationCallback(event, object);
+    };
+    return this._addListener(event, listener);
+  }
+
+  _addListener(event, listener) {
+    targetsAgent.addListener(event, listener);
+    return () => {
+      targetsAgent.removeListener(event, listener);
+    };
+  }
+
+  _toProtocolTarget(target) {
+    const attached = this._sessionsForTarget(target.targetId).length > 0;
+    return Object.assign({ attached }, target);
+  }
+
+  _sessionsForTarget(tid) {
+    return Array.from(this._sessions.entries())
+        .filter(([, { targetId }]) => targetId === tid)
+        .map(([ agentSessionId ]) => agentSessionId);
+  }
+
+  [SessionTerminatedSymbol]() {
+    for (const { detachCallback } of this._sessions.values())
+      detachCallback(() => {});
+    this._removeListeners();
+    this._listeners.forEach((cancel) => cancel());
+  }
+}
+
+class Dispatcher {
+  constructor(session) {
+    this._session = session;
+    this._domainHandlers = new Map();
+  }
+
+  dispatch(message) {
+    if (domainHandlerClasses.size === 0 && this._domainHandlers.size === 0) {
+      return false;
+    }
+
+    if (message === null) {
+      this._shutdown();
+      return;
+    }
+    const parsed = parseSuppressingError(message);
+    const [ domain, method ] = parsed.method ? parsed.method.split('.') : [];
```

Comment:
We shouldn't allow `parsed.method` with more than one dots.

Reply:
Added a simple Regex validation to exclude private methods and such.

Reply:
The regex still allows `parsed.method` with more than two dots, since `method` as an element returned from `.split()` cannot contain a dot. You would have to check `parsed.method.split('.')` directly.

Reply:
I added a check for that scenario.